src/dart/lib/widgets.dart

```diff
@@ -0,0 +1,288 @@
+  State<IconVGImage> createState() => _IconVGImageState();
+}
+
+class _IconVGImageState extends State<IconVGImage> {
+  AssetBundle? _assetBundle;
+  Uint8List? _bytes;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    final AssetBundle assetBundle = DefaultAssetBundle.of(context);
+    if (assetBundle != _assetBundle) {
+      _assetBundle = assetBundle;
+      _bytes = null;
+      _fetchBytes();
+    }
+  }
+
+  @override
+  void didUpdateWidget(IconVGImage oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.asset != widget.asset) {
+      _bytes = null;
+      _fetchBytes();
+    }
+  }
+
+  Future<void> _fetchBytes() async {
+    assert(_assetBundle != null);
+    final ByteData data = await _assetBundle!.load(widget.asset);
+    if (mounted) {
+      setState(() {
+        _bytes = data.buffer.asUint8List();
+      });
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return RawIconVGImage(
+      bytes: _bytes,
+      palette: widget.palette,
+      size: widget.size,
+      clipBehavior: widget.clipBehavior,
+    );
+  }
+}
+
+/// Render the specified [bytes] as an IconVG image.
+///
+/// If the image is an asset, consider using [IconVGImage] instead.
+class RawIconVGImage extends StatefulWidget {
+  const RawIconVGImage({
+    Key? key,
+    this.bytes,
+    this.palette,
+    this.size,
+    this.clipBehavior = Clip.hardEdge,
+  }) : super(key: key);
+
+  /// The bytes. Must be in the IconVG format.
+  final Uint8List? bytes;
+
+  /// The custom color palette used to decode the image.
+  ///
+  /// Must not have more than 64 colors.
+  final List<Color>? palette;
+
+  /// The target render size. Defaults to the [BoxConstraints.maxHeight].
+  final Size? size;
+
+  /// How to clip the image.
+  ///
+  /// This can be used to enable anti-aliased clips if the image is rotated, or
+  /// disable clips entirely (for improved performance) if the image does not
+  /// need them.
+  final Clip clipBehavior;
+
+  @override
+  State<RawIconVGImage> createState() => _RawIconVGImageState();
+}
+
+class _RawIconVGImageState extends State<RawIconVGImage> {
+  IconVGFile? _image;
+  double? _decodedHeight;
+
+  @override
+  void didUpdateWidget(RawIconVGImage oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.bytes != widget.bytes ||
+        oldWidget.palette != widget.palette ||
+        oldWidget.clipBehavior != widget.clipBehavior) {
+        // widget.size is handled in build
+      _image = null;
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (widget.bytes == null) {
+      return SizedBox.fromSize(size: widget.size ?? Size.zero);
+    }
+    return LayoutBuilder(
+      builder: (BuildContext context, BoxConstraints constraints) {
+        final double height = widget.size?.height ?? constraints.maxHeight;
+        if (_image == null || _decodedHeight != height) {
+          try {
+            _image = IconVGFile(
```

Comment:
It would probably help to see the suggestion on changes to `execute` above bubble up here. This call looks cheap (just constructing an object), can actually be quite expensive and probably should never be used in a synchronous build method.

Reply:
A pixel 3 is still significantly more powerful than the average Android device worldwide. That said, these times aren't particularly concerning, but it'd be worth testing something that e.g. draws a gradient across the entirety of the screen.

It'd also be worth looking at what the CPU utilization is - even if it's fast, it's still using more cycles than if you cached this, especially since you know it won't change from frame to frame.

Reply:
Not sure I follow. We are caching the Picture from frame to frame here. What do you want to cache?

I can profile this to see where the expense in the parsing lies. I've done no work to optimize this so far. It's quite possible there's low-hanging fruit.